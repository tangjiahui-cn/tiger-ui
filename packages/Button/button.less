@import '../_theme/variable.less';

@prefix: prefix('btn');
@dangerClass: ~'@{prefix}-danger';
@ghostClass: ~'@{prefix}-ghost';
@loadingClass: ~'@{prefix}-loading';

@{prefix} {
  position: relative;
  box-sizing: border-box;
  font-size: @fontSize;
  background-color: white;
  border-radius: @borderRadius;
  outline: none;
  border: 1px solid @borderColor;
  cursor: pointer;
  transition: all @duration;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  line-height: 1;
  user-select: none;

  &-block {
    width: 100%;
  }

  &-large {
    font-size: @fontSizeLarge;
    height: 40px;
    padding: 10px 20px;
  }

  &-middle {
    height: 32px;
    padding: 8px 16px;
  }

  &-small {
    height: 24px;
    padding: 2px 8px;
  }

  &-loading {
    cursor: default;
  }

  &-primary {
    color: white;
    background: @primary;
    border-color: @primary;

    &:where(@{loadingClass}),
    &:hover {
      background: @primaryHover;
      border-color: @primaryHover;
    }
    &:active {
      background: @primaryActive;
      border-color: @primaryActive;
    }

    &:disabled {
      cursor: not-allowed;
      background: #f5f5f5;
      border-color: #c2c2c2;
      color: #c2c2c2;
    }

    &:where(@{ghostClass}) {
      color: @primary;
      background: transparent;

      &:hover,
      &:focus {
        background: transparent;
      }
    }

    &:where(@{dangerClass}):not(:disabled) {
      background: @danger;
      border-color: @danger;

      &:where(@{loadingClass}),
      &:hover {
        background: @dangerHover;
        border-color: @dangerHover;
      }

      &:active {
        background: @dangerActive;
        border-color: @dangerActive;
      }

      &:where(@{ghostClass}) {
        color: @danger;
        background: transparent;

        &:hover,
        &:focus {
          background: transparent;
        }
      }
    }
  }

  &-default {
    .lineBtn();
  }

  &-dashed {
    .lineBtn(dashed);
  }

  &-dotted {
    .lineBtn(dotted);
  }

  &-text {
    color: @color;
    border-color: transparent;

    &:where(@{loadingClass}),
    &:hover:not(:disabled) {
      background: #ecf5ff;
    }

    &:active:not(:disabled) {
      background: #cce7ff;
    }

    &:disabled {
      cursor: not-allowed;
      background: transparent;
      color: @colorDisabled;
    }

    &:where(@{ghostClass}):not(:disabled) {
      background: transparent;
      color: white;

      &:hover {
        background: transparent;
        color: @primaryHover;
      }
      &:active {
        background: transparent;
        color: @primaryActive;
      }
    }

    &:where(@{dangerClass}):not(:disabled) {
      color: @danger;

      &:where(@{loadingClass}),
      &:hover {
        background: #fdf2f2;
      }

      &:active {
        background: #ffdbdb;
      }

      &:where(@{ghostClass}) {
        background: transparent;
        color: white;
        &:hover {
          background: transparent;
          color: @dangerHover;
        }
        &:active {
          background: transparent;
          color: @dangerActive;
        }
      }
    }
  }
}

.lineBtn (@borderStyle: solid) {
  border-style: @borderStyle;

  &:where(@{loadingClass}),
  &:hover:not(:disabled) {
    color: @primaryHover;
    border-color: @primaryHover;
  }

  &:active:not(:disabled) {
    color: @primaryActive;
    border-color: @primaryActive;
  }

  &:disabled {
    cursor: not-allowed;
    background: #f5f5f5;
    border-color: #c2c2c2;
    color: #c2c2c2;
  }

  &:where(@{ghostClass}) {
    color: white;
    border-color: white;
    background: transparent;
  }

  &:where(@{dangerClass}):not(:disabled) {
    color: @danger;
    border-color: @danger;

    &:where(@{loadingClass}),
    &:hover {
      color: @dangerHover;
      border-color: @dangerHover;
    }

    &:active {
      color: @dangerActive;
      border-color: @dangerActive;
    }

    &:where(@{ghostClass}) {
      color: white;
      border-color: white;
      background: transparent;
    }
  }
}
