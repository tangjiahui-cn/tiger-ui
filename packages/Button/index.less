.my_btn {
  color: #151515;
  background-color: white;
  border: 1px solid #e8e8e8;
  padding: 4px 8px;
  border-radius: 2px;
  cursor: pointer;
  transition: color .2s, border-color .2s;

  &:hover {
    color: blue;
    border-color: blue;
  }
}
