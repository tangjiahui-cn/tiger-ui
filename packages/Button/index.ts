/**
 * Button
 *
 * @author tangjiahui
 * @date 2024/4/13
 */

export { default as Button } from './button';
export type {
  ButtonProps,
  BaseButtonProps,
  ButtonType,
  BaseButtonPropsKeys,
  ButtonSize,
} from './button';
