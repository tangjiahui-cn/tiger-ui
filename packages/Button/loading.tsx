/**
 * Loading
 *
 * @author tangjiahui
 * @date 2024/6/10
 * @description let the button loading effect smoothly.
 */
import React, { useEffect, useState } from 'react';

interface LoadingProps {
  /** loading placeholder width */
  width?: number;
  /** loading effect duration (unit: ms) */
  animationDuration?: number;
  /** loading status  */
  loading?: boolean;
  /** render loading icon.  */
  children?: React.ReactNode;
}

const startDelay = 50; // start loading animation delay.
const hideDelay = 50; // while cancel loading, how time to hide the loading icon.
export default function Loading(props: LoadingProps) {
  const { animationDuration = 400, width = 24, loading, children } = props;
  const [containerWidth, setContainerWidth] = useState(0);
  const [containerVisible, setContainerVisible] = useState(false);
  const [iconVisible, setIconVisible] = useState(false);

  useEffect(() => {
    let containerTimerId: any;
    let iconTimerId: any;

    if (loading) {
      setContainerVisible(true);
      setIconVisible(true);
      setContainerWidth(0);
      containerTimerId = setTimeout(() => {
        setContainerWidth(width);
        containerTimerId = null;
      }, startDelay);
    } else {
      if (width) {
        setContainerWidth(0);
        // if cancel loading, hide icon faster.
        iconTimerId = setTimeout(() => {
          setIconVisible(false);
        }, hideDelay);
        // after animation end, hide the container.
        containerTimerId = setTimeout(() => {
          setContainerVisible(false);
          containerTimerId = null;
        }, animationDuration);
      }
    }

    return () => {
      if (containerTimerId) {
        clearTimeout(containerTimerId);
      }
      if (iconTimerId) {
        clearTimeout(iconTimerId);
      }
    };
  }, [loading]);

  return containerVisible ? (
    <div
      style={{
        width: containerWidth,
        transition: `all ${animationDuration}ms`,
        overflow: 'hidden',
        display: 'inline-flex',
        alignItems: 'center',
        height: 20,
      }}
    >
      {iconVisible && children}
    </div>
  ) : null;
}
