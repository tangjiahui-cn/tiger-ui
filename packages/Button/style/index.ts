import usePrefix from '../../_utils/hooks/usePrefix';
import useToken from '../../_utils/hooks/useToken';
import brotherSelector from '../../_utils/style/brotherSelector';
import type { ButtonType, ButtonSize } from '..';
import useEmotion from '../../_utils/hooks/useEmotion';

type StyleObject = {
  [k: string]: any | StyleObject;
};

export function useStyle(componentName: string): {
  button: (danger?: boolean) => string;
  type: (type: ButtonType) => string;
  size: (type: ButtonSize) => string;
  block: () => string;
  danger: () => string;
  disabled: () => string;
  stayFocus: () => string;
  focus: () => string;
} {
  const token = useToken();
  const prefix = usePrefix(componentName);
  const { css } = useEmotion({ key: prefix });

  const danger = () => `${prefix}-danger`;
  const disabled = () => `${prefix}-disabled`;
  const stayFocus = () => `${prefix}-stay-focus`;
  const focus = () => `${prefix}-focus`;

  function genTypeStyle() {
    const getClassName = (type: ButtonType) => `${prefix}-${type}`;
    const dangerClassName = danger();
    const disabledClassName = disabled();

    const defaultStyle = {
      color: token.color,
      '&:hover': {
        color: token.primaryHover,
        borderColor: token.primaryHover,
      },
      '&:active': {
        color: token.primaryActive,
        borderColor: token.primaryActive,
      },
      // disabled
      [`&.${disabledClassName}`]: {
        cursor: 'not-allowed',
        color: token.colorDisabled,
        borderColor: token.colorDisabled,
        '&:hover': {
          color: token.colorDisabled,
          borderColor: token.colorDisabled,
        },
        '&:active': {
          color: token.colorDisabled,
          borderColor: token.colorDisabled,
        },
      },
      // danger
      [`&.${dangerClassName}`]: {
        color: token.danger,
        borderColor: token.danger,
        '&:hover': {
          color: token.dangerHover,
          borderColor: token.dangerHover,
        },
        '&:active': {
          color: token.dangerActive,
          borderColor: token.dangerActive,
        },
        [`&.${disabledClassName}`]: {
          cursor: 'not-allowed',
          color: token.dangerDisabled,
          borderColor: token.dangerDisabled,
          '&:hover': {
            color: token.dangerDisabled,
            borderColor: token.dangerDisabled,
          },
          '&:active': {
            color: token.dangerDisabled,
            borderColor: token.dangerDisabled,
          },
        },
      },
    };

    const typeStyle = {
      [getClassName('primary')]: {
        color: 'white',
        background: token.primary,
        borderColor: token.primary,
        '&:hover': {
          background: token.primaryHover,
          borderColor: token.primaryHover,
        },
        '&:active': {
          background: token.primaryActive,
          borderColor: token.primaryActive,
        },
        // disabled
        [`&.${disabledClassName}`]: {
          cursor: 'not-allowed',
          background: token.primaryDisabled,
          borderColor: token.primaryDisabled,
          '&:hover': {
            background: token.primaryDisabled,
            borderColor: token.primaryDisabled,
          },
          '&:active': {
            background: token.primaryDisabled,
            borderColor: token.primaryDisabled,
          },
        },
        // danger
        [`&.${dangerClassName}`]: {
          background: token.danger,
          borderColor: token.danger,
          '&:hover': {
            background: token.dangerHover,
            borderColor: token.dangerHover,
          },
          '&:active': {
            background: token.dangerActive,
            borderColor: token.dangerActive,
          },
          [`&.${disabledClassName}`]: {
            cursor: 'not-allowed',
            background: token.dangerDisabled,
            borderColor: token.dangerDisabled,
            '&:hover': {
              background: token.dangerDisabled,
              borderColor: token.dangerDisabled,
            },
            '&:active': {
              background: token.dangerDisabled,
              borderColor: token.dangerDisabled,
            },
          },
        },
      },
      [getClassName('default')]: defaultStyle,
      [getClassName('dashed')]: {
        borderStyle: 'dashed',
        ...defaultStyle,
      },
      [getClassName('dotted')]: {
        borderStyle: 'dotted',
        ...defaultStyle,
      },
      [getClassName('text')]: {
        background: 'white',
        borderColor: 'transparent',
        '&:hover': {
          color: token.colorHover,
        },
        '&:active': {
          color: token.colorActive,
        },
        // disabled
        [`&.${disabledClassName}`]: {
          cursor: 'not-allowed',
          color: token.colorDisabled,
          '&:hover': {
            color: token.colorDisabled,
          },
          '&:active': {
            color: token.colorDisabled,
          },
        },
        // danger
        [`&.${dangerClassName}`]: {
          color: token.danger,
          '&:hover': {
            color: token.dangerHover,
          },
          '&:active': {
            color: token.dangerActive,
          },
          [`&.${disabledClassName}`]: {
            cursor: 'not-allowed',
            color: token.dangerDisabled,
            '&:hover': {
              color: token.dangerDisabled,
            },
            '&:active': {
              color: token.dangerDisabled,
            },
          },
        },
      },
    };

    return {
      type: getClassName,
      typeStyle: typeStyle,
    };
  }

  function genSizeStyle() {
    const getClassName = (size: ButtonSize) => `${prefix}-${size}`;

    const sizeStyle = {
      [getClassName('large')]: {
        padding: '12px 16px',
      },
      [getClassName('middle')]: {
        padding: '8px 14px',
      },
      [getClassName('small')]: {
        padding: '4px 8px',
      },
    };

    return {
      size: getClassName,
      sizeStyle,
    };
  }

  function genBlockStyle() {
    const getClassName = () => `${prefix}-block`;

    const blockStyle = {
      [getClassName()]: {
        width: '100%',
        display: 'block',
      },
    };

    return {
      block: getClassName,
      blockStyle,
    };
  }

  const { type, typeStyle } = genTypeStyle();
  const { size, sizeStyle } = genSizeStyle();
  const { block, blockStyle } = genBlockStyle();

  const button = (danger?: boolean): string => {
    return css({
      fontSize: token.fontSize,
      cursor: 'pointer',
      userSelect: 'none',
      color: token.color,
      background: 'white',
      border: `1px solid ${token.borderColor}`,
      transition: `all ${token.duration}`,
      lineHeight: 1.125,
      verticalAlign: 'middle',

      ...brotherSelector(typeStyle),
      ...brotherSelector(sizeStyle),
      ...brotherSelector(blockStyle),

      [`&:focus.${stayFocus()}`]: {
        borderColor: danger ? token.danger : token.primary,
      },
      [`&.${focus()}`]: {
        borderColor: danger ? token.danger : token.primary,
      },
    });
  };

  return {
    button,
    block,
    type,
    size,
    danger,
    disabled,
    stayFocus,
    focus,
  };
}
