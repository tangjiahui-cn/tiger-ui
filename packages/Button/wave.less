@import '../_theme/variable.less';

@prefix: prefix('wave');
@prefixNoDot: prefixNoDot('wave');
@keyframesShowWave: ~'@{prefixNoDot}-keyframes-show-wave';

@keyframes @keyframesShowWave {
  to {
    box-shadow: 0 0 0 6px transparent;
  }
}

@{prefix} {
  position: absolute;
  left: -1px;
  top: -1px;
  opacity: 0.20;
  border-radius: inherit;
  width: calc(100% + 2px);
  height: calc(100% + 2px);
  pointer-events: none;
  animation: @keyframesShowWave cubic-bezier(0.08, 0.82, 0.17, 1);
  animation-fill-mode: forwards;
}

