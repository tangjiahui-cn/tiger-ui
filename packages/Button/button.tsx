/**
 * Button
 *
 * @author tangjiahui
 * @date 2023/03/16
 */
import React, { DOMAttributes, ForwardedRef } from 'react';
import classNames from 'classnames';
import { SizeType as ButtonSize } from '../_types/common';
import { useGlobal } from '@/_hooks';
import './button.less';
import { omit } from '@/_utils/object';

export type ButtonType = 'primary' | 'dashed' | 'default' | 'text' | 'dotted';
export type { ButtonSize };

export interface BaseButtonProps {
  /**
   * @description 按钮类型
   * @default default
   */
  type?: ButtonType;
  /**
   * @description 按钮禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * @description 是否块级按钮，即宽度100%。
   * @default false
   */
  block?: boolean;
  /**
   * @description 是否危险按钮，危险按钮颜色为醒目色。
   * @default false
   */
  danger?: boolean;
  /**
   * @description 按钮大小，有三种默认尺寸。
   * @default false
   */
  size?: ButtonSize;
  /**
   * @description 保持聚焦（点击后选中样式不会消失，点击其他地方取消选中）
   * @default false
   */
  stayFocus?: boolean;
  /**
   * @description 保持选中（受控状态）
   * @default false
   */
  focus?: boolean;
  /**
   * @description 加载中
   * @default false
   */
  loading?: boolean;
  /**
   * @description style
   */
  style?: React.CSSProperties;
  /**
   * @description className
   */
  className?: string;
}

export type BaseButtonPropsKeys = keyof BaseButtonProps;
export type ButtonProps = BaseButtonProps & DOMAttributes<HTMLButtonElement>;

const privateKeys: BaseButtonPropsKeys[] = [
  'type',
  'disabled',
  'block',
  'danger',
  'size',
  'style',
  'className',
  'stayFocus',
  'focus',
  'loading',
];

const Button: React.ForwardRefExoticComponent<ButtonProps> = React.forwardRef(function (
  props: ButtonProps,
  ref: ForwardedRef<HTMLButtonElement>,
) {
  const { prefix } = useGlobal('btn');
  const originProps: DOMAttributes<HTMLButtonElement> = omit(props, privateKeys);

  const classes = classNames(
    props?.className,
    prefix,
    `${prefix}-${props?.type || 'default'}`,
    `${prefix}-${props?.size || 'middle'}`,
    props?.block && `${prefix}-block`,
    props?.danger && `${prefix}-danger`,
    props?.focus && `${prefix}-focus`,
    props?.stayFocus && `${prefix}-stayFocus`,
  );

  return (
    <button
      {...originProps}
      className={classes}
      ref={ref}
      style={props?.style}
      disabled={props?.disabled}
    >
      {props?.children}
    </button>
  );
});

export default Button;
