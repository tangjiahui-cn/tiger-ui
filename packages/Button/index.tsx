import React, { MouseEventHandler, useRef } from 'react';
import classNames from 'classnames';
import { SizeType as ButtonSize } from '../_types/common';
import { useStyle } from './style';

export type ButtonType = 'primary' | 'dashed' | 'default' | 'text' | 'dotted';
export type { ButtonSize };

export interface ButtonProps {
  /**
   * @description 按钮类型
   * @default default
   */
  type?: ButtonType;
  /**
   * @description 按钮禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * @description 是否块级按钮，即宽度100%。
   * @default false
   */
  block?: boolean;
  /**
   * @description 是否危险按钮，危险按钮颜色为醒目色。
   * @default false
   */
  danger?: boolean;
  /**
   * @description 按钮大小，有三种默认尺寸。
   * @default false
   */
  size?: ButtonSize;
  /**
   * @description 按钮内联样式
   */
  style?: React.CSSProperties;
  /**
   * @description 按钮类class
   */
  className?: string;
  /**
   * @description 保持聚焦（点击后选中样式不会消失，点击其他地方取消选中）
   * @default false
   */
  stayFocus?: boolean;
  /**
   * @description 保持选中（受控状态）
   * @default false
   */
  focus?: boolean;
  /**
   * @description 点击按钮回调事件
   */
  onClick?: MouseEventHandler<HTMLButtonElement>;
  /**
   * @description 子元素
   */
  children?: React.ReactNode;
}

/**
 * 按钮
 *
 * At 2023/04/24
 * By TangJiaHui
 */
export default function Button(props: ButtonProps) {
  const btnRef = useRef(null);
  const style = useStyle('button');

  const classes = classNames([
    style.button(props?.danger),
    style.type(props?.type || 'default'),
    style.size(props?.size || 'middle'),
    props?.block && style.block(),
    props?.danger && style.danger(),
    props?.disabled && style.disabled(),
    props?.stayFocus && style.stayFocus(),
    props?.focus && style.focus(),
    props?.className,
  ]);

  return (
    <button
      ref={btnRef}
      className={classes}
      disabled={props?.disabled}
      onClick={props?.disabled ? undefined : props?.onClick}
      style={props?.style}
    >
      {props?.children}
    </button>
  );
}

Button.defaultProps = {
  type: 'default',
  block: false,
  disabled: false,
  danger: false,
  // loading: false,
  size: 'middle',
  stayFocus: false,
};
