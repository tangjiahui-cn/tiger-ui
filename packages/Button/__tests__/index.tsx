import Button from '..';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  renderTest,
  mountTest,
  nullAndUndefinedChildrenTest,
  reactFragmentTest,
  cssSelectorTest,
} from '../../../tests';
import * as React from 'react';

describe('Button', () => {
  renderTest(Button);
  mountTest(Button);
  mountTest(() => <Button danger>按钮</Button>);
  mountTest(() => <Button type={'primary'}>按钮</Button>);
  mountTest(() => <Button size={'large'}>按钮</Button>);
  mountTest(() => <Button disabled>按钮</Button>);
  mountTest(() => (
    <Button block danger disabled size={'small'} type={'dashed'}>
      按钮
    </Button>
  ));

  describe('dom event test', () => {
    it('onclick trigger', () => {
      const mockClick = jest.fn();
      const wrapper = render(<Button onClick={mockClick} />);
      const dom = wrapper.container.firstChild;
      expect(dom).toBeTruthy();
      if (dom) {
        // @ts-ignore
        fireEvent?.click?.(dom);
        expect(mockClick).toBeCalled();
      }
    });
  });

  cssSelectorTest(['.btn', '.btn-default'], Button);
  cssSelectorTest(['.btn-danger'], () => <Button danger />);
  cssSelectorTest(['.btn-primary'], () => <Button type={'primary'} />);
  cssSelectorTest(['.btn-large'], () => <Button size={'large'} />);
  cssSelectorTest(['.btn-block'], () => <Button block />);
  cssSelectorTest(['[disabled]'], () => <Button disabled />);

  nullAndUndefinedChildrenTest(Button);

  describe('disabled test', () => {
    it('should not onclick when button was disabled.', () => {
      const mockClick = jest.fn();
      const { container } = render(
        <Button disabled onClick={mockClick}>
          按钮
        </Button>,
      );
      const dom = container.firstChild;
      expect(dom).toBeTruthy();
      if (dom) {
        fireEvent.click(dom);
        expect(mockClick).not.toHaveBeenCalled();
      }
    });
  });

  reactFragmentTest(Button);
});
