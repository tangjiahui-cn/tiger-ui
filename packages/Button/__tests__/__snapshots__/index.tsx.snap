// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Null and Undefined Children test should render null or undefined in children without errors 1`] = `
<button
  class="btn btn-default btn-middle"
/>
`;

exports[`Button react fragment test should render React.Fragment without error. 1`] = `
<button
  class="btn btn-default btn-middle"
>
  text
</button>
`;

exports[`Button render test render correctly 1`] = `
<DocumentFragment>
  <button
    class="btn btn-default btn-middle"
  />
</DocumentFragment>
`;
